README.rst

```diff
@@ -1962,6 +1968,347 @@ However, if specifying :code:`Table` objects manually, please make sure to inclu
+
+    graphql_schema = get_schema_with_macros(macro_registry)
+
+Schema for defining macros
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Macro definitions rely on additional directives that are not normally defined in the schema
+the GraphQL compiler uses for querying. We intentionally do not include these directives in
+that schema, since defining macros and writing queries are different modes of use of the compiler,
+and we believe that controlling which sets of directives are available in which mode will minimize
+the potential for user confusion.
+
+You may use the :code:`get_schema_for_macro_definition()` function to transform a querying schema
+into one that is suitable for defining macros.
+
+Macro edges
+~~~~~~~~~~~
+
+Macro edges allow users to define new edges that become part of the GraphQL schema, using existing
+edges as building blocks. They allow users to define shorthand for common querying operations,
+encapsulating uses of existing query functionality (e.g., tags, filters, recursion,
+type coercions, etc.) into a virtual edge with a user-specified name that exists only on a specific
+GraphQL type (and all its subtypes). Both macro edge definitions and their uses are
+fully type-checked, ensuring the soundness of both the macro definition and any queries that use it.
+
+Overview and use of macro edges
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Let us explain the idea of macro edges through a simple example.
+
+Consider the following query, which returns the list of grandchildren of a given animal:
+
+.. code ::
+
+    {
+        Animal {
+            name @filter(op_name: "=", value: ["$animal_name"])
+            out_Animal_ParentOf {
+                out_Animal_ParentOf {
+                    name @output(out_name: "grandchild_name")
+                }
+            }
+        }
+    }
+
+If operations on animals' grandchildren are common in our use case, we may wish that
+an edge like :code:`out_Animal_GrandparentOf` had existed and saved us some repetitive typing.
+
+One of our options is to materialize such an edge in the underlying database itself.
+However, this causes denormalization of the database -- there are now two places where
+an animal's grandchildren are written down -- requiring additional storage space,
+and introducing potential for user confusion and data inconsistency between the two representations.
+
+Another option is to introduce a non-materialized view within the database that *makes it appear*
+that such an edge exists, and query this view via the GraphQL compiler. While this avoids some
+of the drawbacks of the previous approach, not all databases support non-materialized views.
+Also, querying users are not always able to add views to the database, and may require additional
+permissions on the database system.
+
+Macro edges give us the opportunity to define a new :code:`out_Animal_GrandparentOf` edge without
+involving the underlying database systems at all. We simply state that such an edge
```

Comment:
If you follow the suggestion above, please remember to modify this paragraph accordingly.